modules/tensor_mechanics/include/userobjects/CrackFrontPointsProvider.h

```diff
@@ -0,0 +1,41 @@
+/****************************************************************/
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*          All contents are licensed under LGPL V2.1           */
+/*             See LICENSE for full restrictions                */
+/****************************************************************/
+#ifndef CRACKFRONTPOINTSPROVIDER_H
+#define CRACKFRONTPOINTSPROVIDER_H
+
+#include "GeneralUserObject.h"
+
+class CrackFrontPointsProvider;
+
+template <>
+InputParameters validParams<CrackFrontPointsProvider>();
+
+/**
+ * Base class for crack front points provider
+ */
+class CrackFrontPointsProvider : public GeneralUserObject
+{
+public:
+  CrackFrontPointsProvider(const InputParameters & parameters);
+
+  virtual void initialSetup() override{};
+  virtual void initialize() override{};
+  virtual void execute() override{};
+  virtual void finalize() override{};
+
+  /** get a set of points along a crack front from a XFEM GeometricCutUserObject
+   * @return A vector which contains all crack front points
+   */
+  virtual const std::vector<Point>
+  getCrackFrontPoints(unsigned int /*num_crack_front_points*/) const
+  {
+    mooseError("You're trying to use getCrackFrontPoints() for an object that doesn't "
```

Comment:
If you are going to `mooseError` here, did you consider an abstract method instead? That changes a runtime error into a compile time error unless you have objects out there that you need to support. I'd recommend the change unless you have a specific reason not to.

Reply:
Yes, I can make it pure virtual function. Thanks.